src/zenml/zen_server/cache.py

```diff
@@ -0,0 +1,195 @@
+    asynchronous contexts. It uses a simple LRU (Least Recently Used) eviction
+    strategy to manage the cache size.
+
+    Each cache entry has a key, value, timestamp, and expiry. The cache
+    automatically removes expired entries and evicts the oldest entry when
+    the cache reaches its maximum capacity.
+
+
+    Usage Example:
+
+        cache = MemoryCache()
+        uuid_key = UUID("12345678123456781234567812345678")
+
+        cached_or_real_object = cache.get_or_cache(
+            uuid_key, lambda: "sync_data", expiry=120
+        )
+        print(cached_or_real_object)
+    """
+
+    def __init__(self, max_capacity: int, default_expiry: int) -> None:
+        """Initialize the cache with a maximum capacity and default expiry time.
+
+        Args:
+            max_capacity: The maximum number of entries the cache can hold.
+            default_expiry: The default expiry time in seconds.
+        """
+        self.cache: Dict[UUID, MemoryCacheEntry] = OrderedDict()
+        self.max_capacity = max_capacity
+        self.default_expiry = default_expiry
+        self._lock = Lock()
+
+    def set(self, key: UUID, value: Any, expiry: Optional[int] = None) -> None:
+        """Insert value into cache with optional custom expiry time in seconds.
+
+        Args:
+            key: The key to insert the value with.
+            value: The value to insert into the cache.
+            expiry: The expiry time in seconds. If None, uses the default expiry.
+        """
+        with self._lock:
+            self.cache[key] = MemoryCacheEntry(
+                value=value, expiry=expiry or self.default_expiry
+            )
+            self._cleanup()
+
+    def get(self, key: UUID) -> Optional[Any]:
+        """Retrieve value if it's still valid; otherwise, return None.
+
+        Args:
+            key: The key to retrieve the value for.
+
+        Returns:
+            The value if it's still valid; otherwise, None.
+        """
+        with self._lock:
+            return self._get_internal(key)
+
+    def _get_internal(self, key: UUID) -> Optional[Any]:
+        """Helper to retrieve a value without lock (internal use only).
+
+        Args:
+            key: The key to retrieve the value for.
+
+        Returns:
+            The value if it's still valid; otherwise, None.
+        """
+        entry = self.cache.get(key)
+        if entry and not entry.expired:
+            return entry.value
+        elif entry:
+            del self.cache[key]  # Invalidate expired entry
+        return None
+
+    def _cleanup(self) -> None:
+        """Remove expired or excess entries."""
+        # Remove expired entries
+        keys_to_remove = [k for k, v in self.cache.items() if v.expired]
+        for k in keys_to_remove:
+            del self.cache[k]
+
+        # Ensure we don't exceed max capacity
+        while len(self.cache) > self.max_capacity:
+            self.cache.popitem(last=False)  # type: ignore[call-arg]
```

Comment:
Is it possible to annotate `self.cache: OrderedDict[...]` to get rid of this type ignore?

Reply:
I read that `OrderedDict` isn't available as a type annotation in Python 3.8 and decided to use `Dict` instead. But then I realized we don't support Python 3.8 anymore so... yeah, sure :)

Reply:
done